dex/networks/btc/script.go

```diff
@@ -601,6 +627,205 @@ func RefundP2SHContract(contract, sig, pubkey []byte) ([]byte, error) {
+// time must be less than 2^32-1 so that it uses at most 5 bytes. The lockTime
+// is also required to use at least 4 bytes (time stamp, not block time).
+func MakeBondScript(ver uint16, lockTime uint32, pubkeyHash []byte) ([]byte, error) {
+	if ver != 0 {
+		return nil, errors.New("only version 0 bonds supported")
+	}
+	if lockTime >= MaxCLTVScriptNum { // == should be OK, but let's not
+		return nil, errors.New("invalid lock time")
+	}
+	lockTimeInt64 := int64(lockTime)
+	if len(pubkeyHash) != 20 {
+		return nil, errors.New("invalid pubkey hash")
+	}
+	return txscript.NewScriptBuilder().
+		AddInt64(lockTimeInt64).
+		AddOp(txscript.OP_CHECKLOCKTIMEVERIFY).
+		AddOp(txscript.OP_DROP).
+		AddOp(txscript.OP_DUP).
+		AddOp(txscript.OP_HASH160).
+		AddData(pubkeyHash).
+		AddOp(txscript.OP_EQUALVERIFY).
+		AddOp(txscript.OP_CHECKSIG).
+		Script()
+}
+
+// RefundBondScript builds the signature script to refund a time-locked fidelity
+// bond in a P2SH output paying to the provided P2PKH bondScript.
+func RefundBondScript(bondScript, sig, pubkey []byte) ([]byte, error) {
+	return txscript.NewScriptBuilder().
+		AddData(sig).
+		AddData(pubkey).
+		AddData(bondScript).
+		Script()
+}
+
+// RefundBondScript builds the signature script to refund a time-locked fidelity
+// bond in a P2WSH output paying to the provided P2PKH bondScript.
+func RefundBondScriptSegwit(bondScript, sig, pubkey []byte) [][]byte {
+	return [][]byte{
+		sig,
+		pubkey,
+		bondScript,
+	}
+}
+
+// ExtractBondDetailsV0 validates the provided bond redeem script, extracting
+// the lock time and pubkey. The V0 format of the script must be as follows:
+//
+//	<lockTime> OP_CHECKLOCKTIMEVERIFY OP_DROP OP_DUP OP_HASH160 <pubkeyhash[20]> OP_EQUALVERIFY OP_CHECKSIG
+//
+// The script version refers to the pkScript version, not bond version, which
+// pertains to DEX's version of the bond script.
+func ExtractBondDetailsV0(scriptVersion uint16, bondScript []byte) (lockTime uint32, pkh []byte, err error) {
+	type templateMatch struct {
+		expectInt     bool
+		maxIntBytes   int
+		opcode        byte
+		extractedInt  int64
+		extractedData []byte
+	}
+	var template = [...]templateMatch{
+		{expectInt: true, maxIntBytes: 5}, // extractedInt
+		{opcode: txscript.OP_CHECKLOCKTIMEVERIFY},
+		{opcode: txscript.OP_DROP},
+		{opcode: txscript.OP_DUP},
+		{opcode: txscript.OP_HASH160},
+		{opcode: txscript.OP_DATA_20}, // extractedData
+		{opcode: txscript.OP_EQUALVERIFY},
+		{opcode: txscript.OP_CHECKSIG},
+	}
+
+	var templateOffset int
+	tokenizer := txscript.MakeScriptTokenizer(scriptVersion, bondScript)
+	for tokenizer.Next() {
+		if templateOffset >= len(template) {
+			return 0, nil, errors.New("too many script elements")
+		}
+
+		op, data := tokenizer.Opcode(), tokenizer.Data()
+		tplEntry := &template[templateOffset]
+		if tplEntry.expectInt {
+			switch {
+			case data != nil:
+				val, err := dcrtxscript.MakeScriptNum(data, tplEntry.maxIntBytes)
```

Comment:
These "dcrtxscript" functions are identical in the btcd code, but unexported. I'll do a PR in the btcd repo to export them so we can use the correct ones.